src/Control/Director.php

```diff
@@ -566,56 +565,63 @@ public static function absoluteURL($url, $relativeParent = self::BASE)
-     * (Necessary for checking host pre-config)
      * @return string
      */
-    public static function host($respectConfig = true)
+    public static function host()
     {
-        $headerOverride = false;
-        if (TRUSTED_PROXY) {
-            $headers = (getenv('SS_TRUSTED_PROXY_HOST_HEADER')) ? array(getenv('SS_TRUSTED_PROXY_HOST_HEADER')) : null;
-            if (!$headers) {
-                // Backwards compatible defaults
-                $headers = array('HTTP_X_FORWARDED_HOST');
-            }
-            foreach ($headers as $header) {
-                if (!empty($_SERVER[$header])) {
-                    // Get the first host, in case there's multiple separated through commas
-                    $headerOverride = strtok($_SERVER[$header], ',');
-                    break;
-                }
-            }
+        // Check if overridden by alternate_host
+        $host = self::config()->get('alternate_host');
+        if ($host) {
+            return $host;
         }
 
-        if ($respectConfig) {
-            if ($host = Director::config()->uninherited('alternate_host')) {
-                return $host;
+        // Check if overridden by alternate_base_url
+        if ($baseURL = self::config()->get('alternate_base_url')) {
+            $baseURL = self::resolveSubstitutions($baseURL);
+            if (strstr($baseURL, '//')) {
+                return parse_url($baseURL, PHP_URL_HOST);
             }
+        }
 
-            if ($baseURL = Director::config()->uninherited('alternate_base_url')) {
-                if (preg_match('/^(http[^:]*:\/\/[^\/]+)(\/|$)/', $baseURL, $matches)) {
-                    return parse_url($baseURL, PHP_URL_HOST);
+        // Validate proxy-specific headers
+        if (TRUSTED_PROXY) {
+            // Check headers to validate
+            $headers = getenv('SS_TRUSTED_PROXY_HOST_HEADER')
+                ? explode(',', getenv('SS_TRUSTED_PROXY_HOST_HEADER'))
+                : ['HTTP_X_FORWARDED_HOST']; // Backwards compatible defaults
+            foreach ($headers as $header) {
+                if (!empty($_SERVER[$header])) {
+                    // Get the first host, in case there's multiple separated through commas
+                    return strtok($_SERVER[$header], ',');
                 }
             }
         }
 
-        if ($headerOverride) {
-            return $headerOverride;
-        }
-
+        // Check given header
         if (isset($_SERVER['HTTP_HOST'])) {
             return $_SERVER['HTTP_HOST'];
         }
 
-        if ($host = getenv('SS_HOST')) {
-            return $host;
+        // Check base url
+        if ($baseURL = self::config()->uninherited('default_base_url')) {
+            $baseURL = self::resolveSubstitutions($baseURL);
+            if (strstr($baseURL, '//')) {
+                return parse_url($baseURL, PHP_URL_HOST);
+            }
         }
 
+        // Legacy ss4-alpha environment var
+        $legacyHostname = getenv('SS_HOST');
```

Comment:
I disagree; It will mean that users who have just upgraded to alpha6 will have additional pains on this alpha7. We could remove this at stable, but removing it now makes onboarding/upgrading more frustrating to users. I want to make the adoption process as easy and push button for these users as possible. I can mark it as deprecated for removal at 4.0 beta, but I'd rather not immediately break it.